packages/bloc/test/bloc_observer.mocks.dart

```diff
@@ -0,0 +1,24 @@
+import 'package:bloc/bloc.dart';
+import 'package:mockito/mockito.dart';
+
+class MockBlocObserver extends Mock implements BlocObserver {
```

Comment:
You should be able to leave out the forwarding methods: The superclass is `Mock` and that class has a `noSuchMethod` which is not the one from `Object`, and this means that the language automatically induces forwarding methods with the same signature as in the interface, and those forwarding methods will call `noSuchMethod` with a suitable `Invocation` argument. So you'll get code just like the methods below for free.

Reply:
I tried both ways and am still unable to get it working with `Mock`. I am seeing the following error when running tests:

```sh
$ pub run --enable-experiment=non-nullable test
00:01 +4: loading test/bloc_test.dart                                                                                                                                                                                                                 ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:147:23: Warning: Operand of null-aware operation '??' has type 'int' which excludes null.
  int get hashCode => _givenHashCode ?? 0;
                      ^
../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:155:24: Warning: Operand of null-aware operation '??' has type 'String' which excludes null.
  String toString() => _givenName ?? runtimeType.toString();
                       ^
../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:159:10: Warning: Operand of null-aware operation '??' has type 'Iterable<RealCall>' which excludes null.
 - 'Iterable' is from 'dart:core'.
 - 'RealCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
        (realCalls ?? _realCalls).map((call) => call.toString());
         ^
00:01 +4 -1: loading test/bloc_test.dart [E]                                                                                                                                                                                                          
  Failed to load "test/bloc_test.dart":
  Unable to spawn isolate: Error: Cannot run with sound null safety as one or more dependencies do not
  support null safety:
  
   - package:mockito
  
  Run 'pub outdated --mode=null-safety' to determine if versions of your
  dependencies supporting null safety are available.
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:31:11: Error: Field '_whenCall' should be initialized because its type '_WhenCall' doesn't allow null.
   - '_WhenCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
  _WhenCall _whenCall;
            ^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:32:12: Error: Field '_untilCall' should be initialized because its type '_UntilCall' doesn't allow null.
   - '_UntilCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
  _UntilCall _untilCall;
             ^^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:119:61: Error: Optional parameter 'returnValue' should have a default value because its type 'Object' doesn't allow null.
   - 'Object' is from 'dart:core'.
    dynamic noSuchMethod(Invocation invocation, [Object /*?*/ returnValue]) {
                                                              ^^^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:157:49: Error: Optional parameter 'realCalls' should have a default value because its type 'Iterable<RealCall>' doesn't allow null.
   - 'Iterable' is from 'dart:core'.
   - 'RealCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
    String _realCallsToString([Iterable<RealCall> realCalls]) {
                                                  ^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:346:41: Error: Optional parameter 'name' should have a default value because its type 'String' doesn't allow null.
  T named<T extends Mock>(T mock, {String name, int hashCode}) => mock
                                          ^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:346:51: Error: Optional parameter 'hashCode' should have a default value because its type 'int' doesn't allow null.
  T named<T extends Mock>(T mock, {String name, int hashCode}) => mock
                                                    ^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:681:39: Error: Optional parameter 'named' should have a default value because its type 'String' doesn't allow null.
  Null argThat(Matcher matcher, {String named}) =>
                                        ^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:686:43: Error: Optional parameter 'named' should have a default value because its type 'String' doesn't allow null.
  Null captureThat(Matcher matcher, {String named}) =>
                                            ^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:691:43: Error: Optional parameter 'named' should have a default value because its type 'String' doesn't allow null.
  Null typed<T>(ArgMatcher matcher, {String named}) => null;
                                            ^^^^^
../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:147:23: Warning: Operand of null-aware operation '??' has type 'int' which excludes null.
  int get hashCode => _givenHashCode ?? 0;
                      ^
../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:155:24: Warning: Operand of null-aware operation '??' has type 'String' which excludes null.
  String toString() => _givenName ?? runtimeType.toString();
                       ^
../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:159:10: Warning: Operand of null-aware operation '??' has type 'Iterable<RealCall>' which excludes null.
 - 'Iterable' is from 'dart:core'.
 - 'RealCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
        (realCalls ?? _realCalls).map((call) => call.toString());
         ^
00:03 +4 -2: loading test/cubit_test.dart [E]                                                                                                                                                                                                         
  Failed to load "test/cubit_test.dart":
  Unable to spawn isolate: Error: Cannot run with sound null safety as one or more dependencies do not
  support null safety:
  
   - package:mockito
  
  Run 'pub outdated --mode=null-safety' to determine if versions of your
  dependencies supporting null safety are available.
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:31:11: Error: Field '_whenCall' should be initialized because its type '_WhenCall' doesn't allow null.
   - '_WhenCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
  _WhenCall _whenCall;
            ^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:32:12: Error: Field '_untilCall' should be initialized because its type '_UntilCall' doesn't allow null.
   - '_UntilCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
  _UntilCall _untilCall;
             ^^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:119:61: Error: Optional parameter 'returnValue' should have a default value because its type 'Object' doesn't allow null.
   - 'Object' is from 'dart:core'.
    dynamic noSuchMethod(Invocation invocation, [Object /*?*/ returnValue]) {
                                                              ^^^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:157:49: Error: Optional parameter 'realCalls' should have a default value because its type 'Iterable<RealCall>' doesn't allow null.
   - 'Iterable' is from 'dart:core'.
   - 'RealCall' is from 'package:mockito/src/mock.dart' ('../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart').
    String _realCallsToString([Iterable<RealCall> realCalls]) {
                                                  ^^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:346:41: Error: Optional parameter 'name' should have a default value because its type 'String' doesn't allow null.
  T named<T extends Mock>(T mock, {String name, int hashCode}) => mock
                                          ^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:346:51: Error: Optional parameter 'hashCode' should have a default value because its type 'int' doesn't allow null.
  T named<T extends Mock>(T mock, {String name, int hashCode}) => mock
                                                    ^^^^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:681:39: Error: Optional parameter 'named' should have a default value because its type 'String' doesn't allow null.
  Null argThat(Matcher matcher, {String named}) =>
                                        ^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:686:43: Error: Optional parameter 'named' should have a default value because its type 'String' doesn't allow null.
  Null captureThat(Matcher matcher, {String named}) =>
                                            ^^^^^
  ../../../../../../.pub-cache/hosted/pub.dartlang.org/mockito-4.1.2/lib/src/mock.dart:691:43: Error: Optional parameter 'named' should have a default value because its type 'String' doesn't allow null.
  Null typed<T>(ArgMatcher matcher, {String named}) => null;
                                            ^^^^^
00:03 +19 -2: Some tests failed.
```

Do you have any advice/suggestions? I tried the codegen approach but `build_runner` couldn't find any classes to generate. Is this a case where we need to resort to using `Fakes`?

Reply:
First, you need to run the tests with `--no-sound-null-safety` because the test programs import `mockito` which doesn't support null safety yet:

```
> pub run --enable-experiment=non-nullable --no-sound-null-safety test
```

My comment about omitting the forwarding methods was based on properties of the language Dart&mdash;you will indeed get the noSuchMethod forwarders that I mentioned.

However, with mockito the `Mock...` class needs to use a different signature for methods taking parameters with a potentially non-nullable type; and for methods whose return type is potentially non-nullable it needs to call `super.noSuchMethod` with an extra argument. Cf. https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md.

The reason for this is that the setup (`when` and `verify`) relies on passing null as actual arguments, so the mock object methods must permit that. And when the mock object method returns it needs to return a value of the correct type; when the return type is potentially non-nullable it can't use null, and in that case the forwarder must provide some value (any value) of the correct type.

So you get something like the following:

```dart
class MockBlocObserver extends Mock implements BlocObserver {
  @override
  void onChange(Cubit? cubit, Change? change) {
    super.noSuchMethod(Invocation.method(#onChange, [cubit, change]));
  }

  @override
  void onTransition(Bloc? bloc, Transition? transition) {
    super.noSuchMethod(Invocation.method(#onTransition, [bloc, transition]));
  }

  @override
  void onEvent(Bloc? bloc, Object? event) {
    super.noSuchMethod(Invocation.method(#onEvent, [bloc, event]));
  }

  @override
  void onError(Cubit? cubit, Object? error, StackTrace? stackTrace) {
    super.noSuchMethod(Invocation.method(#onError, [cubit, error, stackTrace]));
  }
}
```

The code generation approach is also described on https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md, but manually writing the above will give the same effect.

Reply:
I ended up removing the dependency on `mockito` but what would be the equivalent of

```dart
verifyNever(observer.onError(any, any, any));
````

<img width="1108" alt="Screen Shot 2020-09-18 at 12 11 00 AM" src="https://user-images.githubusercontent.com/8855632/93558610-70dedc00-f943-11ea-9aee-9191fe27b6e7.png">

Reply:
Sounds like you migrated the mockito `Mock` class and declarations like `verifyNever`, or you wrote something which will behave in the same way for your purposes? So `verifyNever` is declared, and it relies on the log data which is similar to that of the `Mock` class to check that `onError` was indeed never called.

If you have a `MockBlockObserver` with an `onError` method which is declared in a library where null safety is enabled, you need to declare that `onError` method such that all parameter types are nullable (so they end in `?` unless it's `void` or `dynamic`, which are nullable already), and similarly for all other methods. With that, you can pass `any` to `onError`, and the logging mechanism should ensure that you get a correct check.

Reply:
Ah that makes sense 👍
Thanks so much for the detailed explanations, I really appreciate it!